packages/kernel-platforms/src/capabilities/fs/shared.ts

```diff
@@ -0,0 +1,118 @@
+import type {
+  PathLike,
+  SyncPathCaveat,
+  ReadFile,
+  Access,
+  ExistsSync,
+  FsConfig,
+  FsCapability,
+} from './types.ts';
+import { fsConfigStruct } from './types.ts';
+import { makeCapabilitySpecification } from '../../specification.ts';
+
+/**
+ * Cross-platform FS operation wrapper with validation (async version)
+ *
+ * @param operation - The underlying operation to wrap
+ * @param pathCaveat - The caveat to apply to path arguments
+ * @returns The operation restricted by the provided caveat
+ */
+export const makeCaveatedFsOperation = <
+  Operation extends (...args: never[]) => unknown,
+>(
+  operation: Operation,
+  pathCaveat: SyncPathCaveat,
+): Operation => {
+  return harden(async (...args: Parameters<Operation>) => {
+    // Assuming first argument is always the path
+    pathCaveat(args[0] as unknown as PathLike);
```

Comment:
```suggestion
    try {
      pathCaveat(args[0] as unknown as PathLike);
    } catch (error) {
      throw new Error(`fs.${operationName}: ${error.message}`, { cause: error });
    }
```

This would produce much more helpful error messages:
`"fs.promises.readFile: Path /etc/passwd is outside allowed root /safe/directory"`
`"fs.existsSync: Symlinks are prohibited: /some/symlink"`

else when path validation fails, the error messages don't provide sufficient context about which operation or which capability triggered the validation failure (readFile, access, existsSync etc)

Reply:
Since the error would arise from the invocation of the capability, there _ought_ to be sufficient context. But I guess `kser` only passes `error.message` back to the kernel. Perhaps this warrants a more general issue.